File: spike/python/src/spike/sources/gitlab.py
# SPDX-License-Identifier: MIT
# Spike PoC — DO NOT EVOLVE INTO PRODUCTION
# Source: .planning/phases/01-language-adr-spike/01-03-PLAN.md
# Locked scope: spike/README.md §Scope
# Pattern reference: .planning/phases/01-language-adr-spike/01-RESEARCH.md §1.1

import asyncio
import os
from typing import Callable

from spike import proc, ui


async def run(target: str, collect: Callable[[str], None]) -> None:
    """Run gitlab-subdomains against target and collect subdomains via text output.

    Source: .planning/phases/01-language-adr-spike/01-RESEARCH.md §1.1 axis (d)
    Parse shape: text output, one subdomain per line.
    Credential: GITLAB_TOKENS env var → file path (per OQ4 locked decision).
    T-01-03-SI-02: token file path passed to subprocess, NOT token contents.
    """
    tokens_path = os.environ.get("GITLAB_TOKENS", "")
    if not tokens_path:
        ui.skip("gitlab-subdomains", "GITLAB_TOKENS unset or empty")
        return
    if not os.path.isfile(tokens_path):
        ui.skip("gitlab-subdomains", f"GITLAB_TOKENS file not found: {tokens_path}")
        return
    if os.path.getsize(tokens_path) == 0:
        ui.skip("gitlab-subdomains", "GITLAB_TOKENS file is empty")
        return

    def line_fn(line: bytes) -> None:
        subdomain = line.decode(errors="replace").strip()
        if subdomain:
            collect(subdomain)

    try:
        await proc.run(
            "gitlab-subdomains",
            ["-d", target, "-t", tokens_path],
            line_fn=line_fn,
        )
    except FileNotFoundError:
        ui.skip("gitlab-subdomains", "not in PATH")
    except asyncio.CancelledError:
        raise
    except Exception as e:
        ui.warn(f"gitlab-subdomains: {e}")

File: spike/go/internal/passive/gitlab.go
// SPDX-License-Identifier: MIT
// Spike PoC — DO NOT EVOLVE INTO PRODUCTION
// Source: .planning/phases/01-language-adr-spike/01-02-PLAN.md Task 3
// Locked scope: spike/README.md §Scope
// Pattern reference: .planning/phases/01-language-adr-spike/01-RESEARCH.md §1.1 (OQ4)
// Bash reference: modules/subdomains.sh:sub_passive() (lines 507-553)
package passive

import (
	"context"
	"os"

	"github.com/six2dez/reconftw/spike/go/internal/proc"
	"github.com/six2dez/reconftw/spike/go/internal/ui"
)

// gitlabRun invokes gitlab-subdomains via proc.Run if GITLAB_TOKENS env var is set.
// Per OQ4 (locked in spike/README.md): reads token file path from env var GITLAB_TOKENS.
// If unset or file empty/unreadable: logs [SKIP] and returns nil.
// T-01-02-SI-02: same token-file-path pattern as githubRun.
func gitlabRun(ctx context.Context, target string, collect func(subdomain, source string)) error {
	tokensPath := os.Getenv("GITLAB_TOKENS")
	if tokensPath == "" {
		ui.Skip("gitlab-subdomains", "GITLAB_TOKENS env var not set")
		return nil
	}

	// Validate token file is readable and non-empty.
	if info, err := os.Stat(tokensPath); err != nil || info.Size() == 0 {
		ui.Skip("gitlab-subdomains", "GITLAB_TOKENS file not found or empty")
		return nil
	}

	ui.Info("gitlab-subdomains: starting for " + target)

	// Command: gitlab-subdomains -d <target> -t <tokens_file>
	args := []string{"-d", target, "-t", tokensPath}

	count := 0
	err := proc.Run(ctx, "gitlab-subdomains", args, func(line []byte) error {
		subdomain := string(line)
		if subdomain != "" {
			collect(subdomain, "gitlab")
			count++
		}
		return nil
	})

	if err != nil {
		ui.Warn("gitlab-subdomains: exited with error: " + err.Error())
		return nil
	}

	ui.Info("gitlab-subdomains: collected " + itoa(count) + " subdomains")
	return nil
}
